packages/core/src/render3/instructions.ts

```diff
@@ -734,19 +735,22 @@ export function directiveCreate<T>(
     }
     previousOrParentNode !.flags = flags;
 
-    ngDevMode && assertDirectivesInRange(index2 - 1);
+    ngDevMode && assertDataInRange(index - 1);
     Object.defineProperty(
         directive, NG_HOST_SYMBOL, {enumerable: false, value: previousOrParentNode});
-    directives[index2] = instance = directive;
-    directives[index2 | 1] = directiveDef;
+    data[index] = instance = directive;
+    index >= ngStaticData.length && (ngStaticData[index] = directiveDef !);
```

Comment:
Use an `if`? It's weird to use a comparator _and then_ an `&&` as a psuedo-if. I would restrict it to the simplest of expressions (e.g. the `ngDevMode` checks), and even then I'm not a huge fan.